File: Backend/src/main/java/com/apartmentservices/models/RealEstateListing.java
package com.apartmentservices.models;

import com.apartmentservices.constant.RealEstateType;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.time.LocalDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
@Entity
@Table(name = "RealEstateListings")
public class RealEstateListing {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ListingID", nullable = false)
    Integer listingId;

    @Column(name = "Title", columnDefinition = "VARCHAR(255) COLLATE utf8mb4_unicode_ci", nullable = false)
    String title;

    @Column(name = "Price", nullable = false)
    Double price;

    @Column(name = "Area", nullable = false)
    Double area; // Diện tích

    @Column(name = "PricePerSquareMeter", nullable = false)
    Double pricePerSquareMeter; // Giá tiền trên 1 m vuông

    @Column(name = "Bedrooms")
    Integer bedrooms;

    @Column(name = "Bathrooms")
    Integer bathrooms;

    @Column(name = "Address", columnDefinition = "VARCHAR(255) COLLATE utf8mb4_unicode_ci", nullable = false)
    String address;

    @Column(name = "DetailedAddress", columnDefinition = "VARCHAR(255) COLLATE utf8mb4_unicode_ci", nullable = false)
    String detailedAddress;

    @Column(name = "Description", columnDefinition = "TEXT")
    String description;

    @ManyToOne
    @JoinColumn(name = "UserID", nullable = false)
    User user; // Người đăng

    @Column(name = "CreatedAt", nullable = false)
    LocalDateTime createdAt;

    @Column(name = "UpdatedAt")
    LocalDateTime updatedAt;

    @Column(name = "ContactPhoneNumber", columnDefinition = "VARCHAR(15) COLLATE utf8mb4_unicode_ci", nullable = false)
    String contactPhoneNumber; // Số điện thoại liên hệ

    @Enumerated(EnumType.STRING)
    @Column(name = "RealEstateType", nullable = false)
    private RealEstateType realEstateType; // Loại bất động sản (Mua bán hay Cho thuê)
}

File: FrontendClient/src/types/RealEstateListing.ts
export type RealEstateListing = {
    listingId: number;
    title: string;
    price: number;
    area: number;
    pricePerSquareMeter: number;
    bedrooms: number;
    bathrooms: number;
    address: string;
    detailedAddress: string | null;
    description: string;
    contactPhoneNumber: string;
    createdAt: string;
    updatedAt: string | null;
    mediaList: RealEstateMedia
    user: User
};


export type RealEstateMedia = {
    map(arg0: (media: import("../redux/thunks/realestate").EstateMedia) => string): string[];
    mediaId: number;
    mediaUrl: string;
    mediaType: string;
    seq: number;
};

interface User {
    id: null;
    username: string;
    firstName: string;
    lastName: string;
    avatar: string;
    roles: Role[];
  }
  
  interface Role {
    name: string;
    description: string;
    permissions: any[];
  }